src/dbus.c

```diff
@@ -439,6 +445,145 @@ static void dbus_cb_dunst_NotificationRemoveFromHistory(GDBusConnection *connect
+                                              "alignment",
+                                              "s",
+                                              enum_to_string(horizontal_alignment_enum_data, r->alignment));
+                if (r->hide_text != -1)
+                        g_variant_dict_insert(&dict, "hide_text", "b", r->hide_text);
+                if (r->progress_bar_alignment != -1)
+                        g_variant_dict_insert(&dict,
+                                              "progress_bar_alignment",
+                                              "s",
+                                              enum_to_string(horizontal_alignment_enum_data,
+                                                             r->progress_bar_alignment));
+                if (r->min_icon_size != -1)
+                        g_variant_dict_insert(&dict, "min_icon_size", "i", r->min_icon_size);
+                if (r->max_icon_size != -1)
+                        g_variant_dict_insert(&dict, "max_icon_size", "i", r->max_icon_size);
+                if (r->action_name)
+                        g_variant_dict_insert(&dict, "action_name", "s", r->action_name);
+                if (r->set_category)
+                        g_variant_dict_insert(&dict, "set_category", "s", r->set_category);
+                if (r->markup != MARKUP_NULL)
+                        g_variant_dict_insert(&dict, "markup", "s", enum_to_string(markup_mode_enum_data, r->markup));
+                if (r->icon_position != -1)
+                        g_variant_dict_insert(&dict,
+                                              "icon_position",
+                                              "s",
+                                              enum_to_string(icon_position_enum_data, r->icon_position));
+                if (r->fg)
+                        g_variant_dict_insert(&dict, "fg", "s", r->fg);
+                if (r->bg)
+                        g_variant_dict_insert(&dict, "bg", "s", r->bg);
+                if (r->highlight)
+                        g_variant_dict_insert(&dict, "highlight", "s", r->highlight);
+                if (r->fc)
+                        g_variant_dict_insert(&dict, "fc", "s", r->fc);
+                if (r->format)
+                        g_variant_dict_insert(&dict, "format", "s", r->format);
+                if (r->default_icon)
+                        g_variant_dict_insert(&dict, "default_icon", "s", r->default_icon);
+                if (r->new_icon)
+                        g_variant_dict_insert(&dict, "new_icon", "s", r->new_icon);
+                if (r->script)
+                        g_variant_dict_insert(&dict, "script", "s", r->script);
+                if (r->set_stack_tag)
+                        g_variant_dict_insert(&dict, "set_stack_tag", "s", r->set_stack_tag);
+                if (r->override_pause_level != -1)
+                        g_variant_dict_insert(&dict, "override_pause_level", "i", r->override_pause_level);
```

Comment:
In this function, is it not possible to use the data in settings_data.h to programmatically fill the dictionary? All rules in the settings data struct have a non-zero rule (except probably for the first one, if that's a problem that's fixable).
This would make it more easy to maintain. Otherwise it's easy to forget to add it here after adding a rule. If you have idea's for how to do this, please let me know. Then we can discuss if it's worth doing.

Reply:
My knowledge of C is rather limited, but if I got you right, this would mean compile time code generation, which would mean using macros. I'm not 100% sure on this, but I would highly doubt, that this is possible. Run-time code generation via reflection is not possible in C AFAIK.

My (maybe overly simplistic) approach: just write some comment to the definition of rule to remind the implementer to add it to `dbus_cb_dunst_RuleList` as well.

Reply:
Okay, could you then add some documentation to settings_data.h?

Reply:
maybe I can make it automatic when I refactor rule parsing. for now it is fine though